In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

%matplotlib inline

In [2]:
df = pd.read_csv('data.csv')
df

order_id  shop_id  user_id  order_amount  total_items payment_method  \
0            1       53      746           224            2           cash   
1            2       92      925            90            1           cash   
2            3       44      861           144            1           cash   
3            4       18      935           156            1    credit_card   
4            5       18      883           156            1    credit_card   
...        ...      ...      ...           ...          ...            ...   
4995      4996       73      993           330            2          debit   
4996      4997       48      789           234            2           cash   
4997      4998       56      867           351            3           cash   
4998      4999       60      825           354            2    credit_card   
4999      5000       44      734           288            2          debit   

               created_at  
0     2017-03-13 12:36:56  
1     2017-03-03 17:38:52  
2      2017-03-14 4:23:56  
3     2017-03-26 12:43:37  
4      2017-03-01 4:35:11  
...                   ...  
4995  2017-03-30 13:47:17  
4996  2017-03-16 20:36:16  
4997   2017-03-19 5:42:42  
4998  2017-03-16 14:51:18  
4999  2017-03-18 15:48:18  

[5000 rows x 7 columns]

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 7 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   order_id        5000 non-null   int64 
 1   shop_id         5000 non-null   int64 
 2   user_id         5000 non-null   int64 
 3   order_amount    5000 non-null   int64 
 4   total_items     5000 non-null   int64 
 5   payment_method  5000 non-null   object
 6   created_at      5000 non-null   object
dtypes: int64(5), object(2)
memory usage: 273.6+ KB


Note: No Null attributes identified on numerical features

In [4]:
df.describe()

order_id      shop_id      user_id   order_amount  total_items
count  5000.000000  5000.000000  5000.000000    5000.000000   5000.00000
mean   2500.500000    50.078800   849.092400    3145.128000      8.78720
std    1443.520003    29.006118    87.798982   41282.539349    116.32032
min       1.000000     1.000000   607.000000      90.000000      1.00000
25%    1250.750000    24.000000   775.000000     163.000000      1.00000
50%    2500.500000    50.000000   849.000000     284.000000      2.00000
75%    3750.250000    75.000000   925.000000     390.000000      3.00000
max    5000.000000   100.000000   999.000000  704000.000000   2000.00000

In [5]:
total_amount = df['order_amount'].sum()
total_items = df['total_items'].sum()
total_count_items = df['order_amount'].count() #df['total_items'].count()

AOV_rectified = total_amount/total_items
print(AOV_rectified)

AOV_reported = total_amount/total_count_items
print(AOV_reported)

357.92152221412965
3145.128


## For First Question Part - A

#### ANSWER: 
Looking at the data above, The actual mean of `order_amount -$375.921`. According to shopify data, the naively calculated AOV was $3145.13. Instead of calculating the sum of 'total_items', employee has calculated the total count of 'total_items' - resulting in false AOV.

Looking at AOV_rectified & AOV_reported, we can clearly see the difference. Along with that, df.describe() shows the calculated mean as `3145.128000`. And the calculation behind that was clearly same as AOV_reported.

## For First Question Part-B

#### ANSWER: 
For shopping data such as Shopify, If we want to calculate the AOV of the sneaker shops - better approch can be to look at individual shops data & clearly understand the total purchases by an individual shop & follow a metrics like mean, median, mode. Later we use MMR(mode to mean ratio) if mode & mean has high difference.

In [6]:
# Before going into that let's see of the given data is actual for 30-days window
df.sort_values(by='created_at')

order_id  shop_id  user_id  order_amount  total_items payment_method  \
1862      1863       39      738           536            4           cash   
1741      1742       39      910           268            2           cash   
3228      3229       97      912           324            2           cash   
1267      1268       80      798           290            2    credit_card   
2689      2690       49      799           258            2    credit_card   
...        ...      ...      ...           ...          ...            ...   
2765      2766        9      708           236            2          debit   
4890      4891       63      853           136            1           cash   
244        245       32      928           202            2           cash   
211        212       22      712           292            2    credit_card   
145        146       72      906           640            4           cash   

              created_at  
1862  2017-03-01 0:08:09  
1741  2017-03-01 0:10:19  
3228  2017-03-01 0:14:12  
1267  2017-03-01 0:19:31  
2689  2017-03-01 0:22:25  
...                  ...  
2765  2017-03-30 9:22:41  
4890  2017-03-30 9:27:00  
244   2017-03-30 9:30:28  
211   2017-03-30 9:40:40  
145   2017-03-30 9:55:00  

[5000 rows x 7 columns]

Given data is clearly for 30-day window

In [7]:
data_copy = df.copy()

In [8]:
# Calculating the total purchases amount, total items by individual shop_id
total_items_by_shopID = df.groupby(['shop_id'])['total_items'].sum()
total_amount_by_shopID = df.groupby(['shop_id'])['order_amount'].sum()
new_df = pd.concat([total_items_by_shopID,total_amount_by_shopID], axis=1).reset_index(level=0)
new_df['AOV'] = new_df['order_amount']/new_df['total_items']
new_df

shop_id  total_items  order_amount    AOV
0         1           86         13588  158.0
1         2          102          9588   94.0
2         3           99         14652  148.0
3         4          103         13184  128.0
4         5           92         13064  142.0
..      ...          ...           ...    ...
95       96          110         16830  153.0
96       97           96         15552  162.0
97       98          107         14231  133.0
98       99           94         18330  195.0
99      100           77          8547  111.0

[100 rows x 4 columns]

In [9]:
new_df.describe()

shop_id   total_items  order_amount           AOV
count  100.000000    100.000000  1.000000e+02    100.000000
mean    50.500000    439.360000  1.572564e+05    407.990000
std     29.011492   3396.366111  1.216218e+06   2557.462906
min      1.000000     67.000000  6.840000e+03     90.000000
25%     25.750000     88.000000  1.293050e+04    132.750000
50%     50.500000    100.000000  1.488750e+04    153.000000
75%     75.250000    111.250000  1.760000e+04    168.250000
max    100.000000  34063.000000  1.199018e+07  25725.000000

In [10]:
print('Mean:',new_df['AOV'].mean())
print('Median:',new_df['AOV'].median())
print('Mode:',new_df['AOV'].mode())

Mean: 407.99
Median: 153.0
Mode: 0    153.0
dtype: float64


Looking at the above metrics, mode & median is same for this data. But mean appears to be higher than median. This usual happens due to data anaomly. Let's have a look at our main df & new_df again to get a clear picture.

In [11]:
new_df.sort_values(by='order_amount',ascending=False)

shop_id  total_items  order_amount      AOV
41       42        34063      11990176    352.0
77       78           88       2263800  25725.0
88       89          118         23128    196.0
80       81          128         22656    177.0
5         6          121         22627    187.0
..      ...          ...           ...      ...
1         2          102          9588     94.0
99      100           77          8547    111.0
55       56           69          8073    117.0
31       32           79          7979    101.0
91       92           76          6840     90.0

[100 rows x 4 columns]

Based on above results, `Shop_id['42']` has higher purchase history with `total items` = 3406, `order_amount` = $11990176
`shop_id['78']` has purchase history of total_items = 88 & order_amount = 2263800. And AOV of `shop_id['78']` looks unusal. Lets have a look at original data.

In [12]:
df[df['shop_id']==78] 

order_id  shop_id  user_id  order_amount  total_items payment_method  \
160        161       78      990         25725            1    credit_card   
490        491       78      936         51450            2          debit   
493        494       78      983         51450            2           cash   
511        512       78      967         51450            2           cash   
617        618       78      760         51450            2           cash   
691        692       78      878        154350            6          debit   
1056      1057       78      800         25725            1          debit   
1193      1194       78      944         25725            1          debit   
1204      1205       78      970         25725            1    credit_card   
1259      1260       78      775         77175            3    credit_card   
1384      1385       78      867         25725            1           cash   
1419      1420       78      912         25725            1           cash   
1452      1453       78      812         25725            1    credit_card   
1529      1530       78      810         51450            2           cash   
2270      2271       78      855         25725            1    credit_card   
2452      2453       78      709         51450            2           cash   
2492      2493       78      834        102900            4          debit   
2495      2496       78      707         51450            2           cash   
2512      2513       78      935         51450            2          debit   
2548      2549       78      861         25725            1           cash   
2564      2565       78      915         77175            3          debit   
2690      2691       78      962         77175            3          debit   
2773      2774       78      890         25725            1           cash   
2818      2819       78      869         51450            2          debit   
2821      2822       78      814         51450            2           cash   
2906      2907       78      817         77175            3          debit   
2922      2923       78      740         25725            1          debit   
3085      3086       78      910         25725            1           cash   
3101      3102       78      855         51450            2    credit_card   
3151      3152       78      745         25725            1    credit_card   
3167      3168       78      927         51450            2           cash   
3403      3404       78      928         77175            3          debit   
3440      3441       78      982         25725            1          debit   
3705      3706       78      828         51450            2    credit_card   
3724      3725       78      766         77175            3    credit_card   
3780      3781       78      889         25725            1           cash   
4040      4041       78      852         25725            1           cash   
4079      4080       78      946         51450            2           cash   
4192      4193       78      787         77175            3    credit_card   
4311      4312       78      960         51450            2          debit   
4412      4413       78      756         51450            2          debit   
4420      4421       78      969         77175            3          debit   
4505      4506       78      866         25725            1          debit   
4584      4585       78      997         25725            1           cash   
4715      4716       78      818         77175            3          debit   
4918      4919       78      823         25725            1           cash   

               created_at  
160    2017-03-12 5:56:57  
490   2017-03-26 17:08:19  
493   2017-03-16 21:39:35  
511    2017-03-09 7:23:14  
617   2017-03-18 11:18:42  
691   2017-03-27 22:51:43  
1056  2017-03-15 10:16:45  
1193  2017-03-16 16:38:26  
1204  2017-03-17 22:32:21  
1259   2017-03-27 9:27:20  
1384  2017-03-17 16:38:06  
141

In [13]:
df[df['shop_id']==42] 

order_id  shop_id  user_id  order_amount  total_items payment_method  \
15          16       42      607        704000         2000    credit_card   
40          41       42      793           352            1    credit_card   
60          61       42      607        704000         2000    credit_card   
308        309       42      770           352            1    credit_card   
409        410       42      904           704            2    credit_card   
520        521       42      607        704000         2000    credit_card   
834        835       42      792           352            1           cash   
835        836       42      819           704            2           cash   
938        939       42      808          1056            3    credit_card   
979        980       42      744           352            1          debit   
1104      1105       42      607        704000         2000    credit_card   
1362      1363       42      607        704000         2000    credit_card   
1364      1365       42      797          1760            5           cash   
1367      1368       42      926          1408            4           cash   
1436      1437       42      607        704000         2000    credit_card   
1471      1472       42      907          1408            4          debit   
1512      1513       42      946           352            1          debit   
1520      1521       42      756           704            2          debit   
1562      1563       42      607        704000         2000    credit_card   
1602      1603       42      607        704000         2000    credit_card   
1911      1912       42      739           704            2           cash   
1929      1930       42      770           352            1    credit_card   
2003      2004       42      934           704            2           cash   
2018      2019       42      739           352            1          debit   
2053      2054       42      951           352            1          debit   
2153      2154       42      607        704000         2000    credit_card   
2273      2274       42      747           704            2          debit   
2297      2298       42      607        704000         2000    credit_card   
2491      2492       42      868           704            2          debit   
2609      2610       42      868           704            2          debit   
2766      2767       42      970           704            2    credit_card   
2835      2836       42      607        704000         2000    credit_card   
2969      2970       42      607        704000         2000    credit_card   
2987      2988       42      819          1056            3           cash   
3332      3333       42      607        704000         2000    credit_card   
3513      3514       42      726          1056            3          debit   
3651      3652       42      830           352            1    credit_card   
3697      3698       42      839           352            1          debit   
3903      3904       42      975           352            1          debit   
3998      3999       42      886           352            1          debit   
4056      4057       42      607        704000         2000    credit_card   
4231      4232       42      962           352            1           cash   
4294      4295       42      859           704            2           cash   
4326      4327       42      788           704            2          debit   
4421      4422       42      736           704            2    credit_card   
4625      4626       42      809           352            1    credit_card   
4646      4647       42      607        704000         2000    credit_card   
4745      4746       42      872           352            1          debit   
4767      4768       42      720           704            2    credit_card   
4868      4869       42      607        704000         2000    credit_card   
4882      4883       42     

By comparing the both `shop_id[42,78]`, the individual item cost of each order appears to be constant. Assume that each shop sells only one model, the prices of these models may change. And `shop_id['42']` appears to be selling high-end model & price of a shoe is `$352` - Looks geniune. 

Both for `shop_id['78']`, the individual item cost of each order is `$25725`. And this appears to be a mistake in data entry or a fraudulent data. Due to unusual pattern of this data, we cannot use this entry_point for AOV calculation till we get a standard report. As a result, by assuming this data as an anaomly - we're dropping `shop_id['78']` data.

In [14]:
#Remove data anaomly
new_df.drop(index=77, inplace=True)

In [15]:
new_df

shop_id  total_items  order_amount    AOV
0         1           86         13588  158.0
1         2          102          9588   94.0
2         3           99         14652  148.0
3         4          103         13184  128.0
4         5           92         13064  142.0
..      ...          ...           ...    ...
95       96          110         16830  153.0
96       97           96         15552  162.0
97       98          107         14231  133.0
98       99           94         18330  195.0
99      100           77          8547  111.0

[99 rows x 4 columns]

In [17]:
Mean = new_df['AOV'].mean()
Median = new_df['AOV'].median()
Mode = new_df['AOV'].mode()
print('Mean:',Mean)
print('Median:',Median)
print('Mode:',Mode)

Mean: 152.26262626262627
Median: 153.0
Mode: 0    153.0
dtype: float64


In [19]:
MMR = Mode/Mean*100
MMR

# If MMR is higher, mean is our AOV & if MMR is lower than 67%, we can use mode as our AOV.

0    100.484278
dtype: float64

## For First Question Part-C

### Answer
By eliminating the anamoly from our data, mode & median appear to be same. As a result, we conclude that 50% of customers spent equal or less than & other 50% of cutomers spent equal or more than `$153.00`. And based on Mode to mean ratio calculation, our AOV is `$152.2`.

# Second Question

Question 2: For this question you’ll need to use SQL. Follow this link to access the data set required for the challenge. Please use queries to answer the following questions. Paste your queries along with your final numerical answers below.

### 1.How many orders were shipped by Speedy Express in total?
	Ans: 54

`   SELECT COUNT(DISTINCT orderID) as total_orders
    FROM Orders
    JOIN Shippers
    ON Orders.ShipperID = Shippers.ShipperID
    WHERE Shippers.ShipperName = 'Speedy Express' `


### 2.What is the last name of the employee with the most orders?

Ans: Peacock, 40

`    SELECT e.LastName, COUNT(DISTINCT o.OrderID) as total_count
    FROM Orders o
    JOIN Employees e
    ON o.EmployeeID = e.employeeID
    GROUP BY o.EmployeeID
    ORDER BY total_count desc
    LIMIT 1 `





### 3.What product was ordered the most by customers in Germany?

Ans: ‘Boston Crab Meat’ orders: 4, 160 (In-terms of quantity)

‘Gorgonzola Telino’, 5, 125 (In-terms of total distinct orders) 

`   SELECT p.ProductName, COUNT(DISTINCT od.OrderID) as orders, SUM(od.Quantity) as quantity
    FROM Orders o
    JOIN OrderDetails od
    ON od.OrderID = o.OrderID
    JOIN Customers c
    ON o.CustomerID = c.CustomerID
    JOIN Products p
    ON od.ProductID = p.ProductID
    WHERE c.Country = 'Germany'
    GROUP BY p.ProductID
    ORDER BY quantity desc  `
